core/src/repair_weight.rs

```diff
@@ -249,28 +271,54 @@ impl RepairWeight {
     }
 
     /// Split `slot` and descendants into an orphan tree in repair weighting
-    /// These orphaned slots should be removed from `unrooted_slots` as on proper repair these slots might
-    /// now be part of the rooted path
+    /// Even if these orphaned slots were previously pruned, they should be added back properly as
+    /// on repair they might now be part of the rooted path.
+    /// Assumes that `slot` is greater than `self.root`
     pub fn split_off(&mut self, slot: Slot) {
+        assert!(slot >= self.root);
         if slot == self.root {
             error!("Trying to orphan root of repair tree {}", slot);
             return;
         }
-        if let Some(subtree_root) = self.slot_to_tree.get(&slot) {
+        if let Some(TreeRoot::Root(subtree_root)) = self.slot_to_tree.get(&slot) {
             if *subtree_root == slot {
                 info!("{} is already orphan, skipping", slot);
                 return;
             }
             let subtree = self
                 .trees
                 .get_mut(subtree_root)
-                .expect("subtree must exist");
+                .expect("`self.slot_to_tree` and `self.trees` must be in sync");
             let orphaned_tree = subtree.split_off(&(slot, Hash::default()));
-            for ((orphaned_slot, _), _) in orphaned_tree.all_slots_stake_voted_subtree() {
-                self.unrooted_slots.remove(orphaned_slot);
-                self.slot_to_tree.insert(*orphaned_slot, slot);
-            }
+            self.rename_tree_root(&orphaned_tree, TreeRoot::Root(slot));
             self.trees.insert(slot, orphaned_tree);
```

Comment:
This logic is also shared in the `PrunedRoot` branch below, can maybe be moved to a `fn split_into_orphan_tree()` which can then be unit tested